Add the SpawnStarted signal

refi64 opened the pull request:
Now we can see PIDs of spawned processes...but there's no way to actually know what spawned processes they correspond to. This adds the ability to see that.

alexlarsson commented on data/org.freedesktop.portal.Flatpak.xml:
```diff
@@ -228,6 +228,27 @@
     </method>
 
+    <!--
+        SpawnRelativePid:
```
This name looks like it spawns something. I think a better name would be something like GetSpawnedPid or something.

alexlarsson commented on portal/flatpak-portal.c:
```diff
@@ -1024,6 +1026,124 @@ handle_spawn_signal (PortalFlatpak         *object,
+
+          return TRUE;
+        }
+    }
```
Iterating over all instances like this seems quite expensive and fragile. I think a better option would be to add an option to flatpak run to write the instance it to a file descriptor, and then record this in the PidData.

refi64 replied:
Hmm my original thought was actually to add an argument that corresponds to bwrap's `--json-status-fd`, then just emit a `SpawnSuccess` signal with the PID relative to the spawner's namespace, which would be a bit prettier overall. Would that work? I just wasn't certain about the upstream feel for adding fd-related arguments to `flatpak run` itself...

refi64 replied:
FWIW in terms of the use case (Chromium) this would work just as well.

refi64 commented:
Okay I completely redid the way it all works, now instead it emits a `SpawnStarted` signal once the process is fully started and passes the relative PID along.

alexlarsson commented on portal/flatpak-portal.c:
```diff
@@ -844,6 +1064,35 @@ handle_spawn (PortalFlatpak         *object,
+                                 instance_id_read_finish, instance_id_read_data);
+
+      g_ptr_array_add (flatpak_argv, g_strdup_printf ("--instance-id-fd=%d", fds[1]));
+      child_setup_data.instance_id_fd = fds[1];
```
This should use flatpak_bwrap_add_fd() instead. It will automatically inherit the fd and close it when not needed anymore (so also no need to do the output stream workaround above.

alexlarsson commented on portal/flatpak-portal.c:
```diff
@@ -844,6 +1064,35 @@ handle_spawn (PortalFlatpak         *object,
+      instance_id_out_stream = G_OUTPUT_STREAM (g_unix_output_stream_new (fds[1], TRUE));
+
+      instance_id_read_data = g_new0 (InstanceIdReadData, 1);
+      instance_id_read_data->buffer = g_malloc0 (INSTANCE_ID_BUFFER_SIZE);
```
Why don't you just put the buffer inline in the InstanceIdReadData struct? No need for a separate allocation.

alexlarsson commented on portal/flatpak-portal.c:
```diff
@@ -763,7 +983,7 @@ handle_spawn (PortalFlatpak         *object,
     }
 
-  g_ptr_array_add (flatpak_argv, g_strdup ("flatpak"));
+  g_ptr_array_add (flatpak_argv, g_strdup ("./flatpak"));
```
This should not be in the final commit

alexlarsson commented on portal/flatpak-portal.c:
```diff
@@ -247,10 +251,218 @@ typedef struct
+  g_signal_connect (monitor, "changed", G_CALLBACK (bwrapinfo_changed), monitor_data);
+
+  check_child_pid_status (monitor, monitor_data);
+}
```
This seems racy in general. There is no guarantee that we'll reach this code before the file is written, and in that case we would never get any change event. Also, file notifiers like this are super complicated internally and can run out of inotify handles and whatnot. I would recommend to just do the check, then if its no there yet, just do a loop via an timeout instead of the file monitor.

refi64 replied:
Do note that in flatpak-run.c, the message is sent to the fd *after* bwrapinfo.json is created, so it should at least exist at this point. That being said if you prefer a timeout I can do that, would 100ms be fine? (I don't want to make it too large, since if you're spawning several processes it could cause visible delays.)

alexlarsson replied:
Knowing that it exists is perhaps nice, but the problem is that it could be ready already and never see any change so will never trigger the read. 
I think what you want to do is immediately try to read the file, then do a say 1msec timeout and after that back it down to say 100msec.

refi64 commented:
Okay so *finally* should be done again, the build with old glib test seems to be an unrelated issue with setting up the testing environment.

One question:
> This should use flatpak_bwrap_add_fd() instead. It will automatically inherit the fd and close it when not needed anymore (so also no need to do the output stream workaround above.

I'm admittedly a tad confused as to how this work work. The output stream workaround is to keep the fd alive on the portal's end, so if the spawn fails FDs don't get leaked. `flatpak_bwrap_add_fd` requires a FlatpakBwrap instance, which is only available inside `flatpak-run.c` which would be post-fork-and-exec, at which point the portal's workaround doesn't even matter anymore. Am I misunderstanding something?

refi64 commented:
Pushed a fix for a nasty pointer deref issue & should fix the old glib build test.

alexlarsson commented on app/flatpak-builtins-build.c:
```diff
@@ -537,7 +537,7 @@ flatpak_builtin_build (int argc, char **argv, GCancellable *cancellable, GError
                                       app_id_dir, app_context, NULL,
                                       FALSE, TRUE, TRUE,
-                                      &app_info_path,
+                                      &app_info_path, 0,
```
Should use -1 for no fd.

alexlarsson commented on common/flatpak-installation.c:
```diff
@@ -709,7 +709,7 @@ flatpak_installation_launch_full (FlatpakInstallation *self,
                         NULL,
                         NULL,
-                        NULL, 0,
+                        NULL, 0, 0,
```
should use -1 for no fd

alexlarsson commented on portal/flatpak-portal.c:
```diff
@@ -247,10 +251,188 @@ typedef struct
+  if (pid_data == NULL)
+    {
+      g_warning ("%u already exited, skipping SpawnStarted", pid);
+      return G_SOURCE_REMOVE;
```
Shouldn't we do this check before the timeout above, otherwise we will never stop polling with timeouts if the child has exited.

alexlarsson commented on portal/flatpak-portal.c:
```diff
@@ -247,10 +251,188 @@ typedef struct
+      g_autoptr(GError) error = NULL;
+      relative_child_pid = get_child_pid_relative_to_parent_sandbox (child_pid, &error);
+      if (relative_child_pid == 0)
+        g_error ("Failed to find relative PID for %d: %s", child_pid, error->message);
```
g_error is a fatal operation that will kill the portal process, i don't thnk we want that.

refi64 commented:
@alexlarsson 

> should use -1 for no fd

I changed that a while back, I think you're looking at old commits :sweat_smile: Fixed up everything else that was mentioned.

alexlarsson commented:
> > should use -1 for no fd
> 
> I changed that a while back, I think you're looking at old commits sweat_smile Fixed up everything else that was mentioned.

I noticed that they were fixed in a later commit.

alexlarsson commented on portal/flatpak-portal.c:
```diff
@@ -247,10 +251,195 @@ typedef struct
+    {
+      /* 0 means EOF, so the process could never have been started. */
+      if (bytes_read == -1)
+        g_error ("Failed to read instance id: %s", error->message);
```
Here too we shouldn't use g_error (as it will kill the portal).

alexlarsson commented:
> > This should use flatpak_bwrap_add_fd() instead. It will automatically inherit the fd and close it when not needed anymore (so also no need to do the output stream workaround above.
> 
> I'm admittedly a tad confused as to how this work work. The output stream workaround is to keep the fd alive on the portal's end, so if the spawn fails FDs don't get leaked. `flatpak_bwrap_add_fd` requires a FlatpakBwrap instance, which is only available inside `flatpak-run.c` which would be post-fork-and-exec, at which point the portal's workaround doesn't even matter anymore. Am I misunderstanding something?

Ah, I was confused about the availability of FlatpakBwrap when reading this. But this looks ok.
